Lotto game PR

gkfla668 opened the pull request:
오빠 처음으로 mvc 패턴 적용해서 처음부터 코드 구현하려니.. 이게 맞는건지 잘 감이 잡히지 않아서 😅 
자동차경주게임 다른 분들이 mvc 패턴 적용한거 참고하면서 얼추 알거같으면서도 model 부분이 어렵고 나머지 부분도 아직 잘 모르겠네요 ㅠㅠ 구조만 분리해보고 피드백 받고자 당첨 통계는 아직 구현 못하고 pr 요청 합니다 !!

gkfla668 committed:
feat: 추가 InputView, OutputView

gkfla668 committed:
refactor: 수정 Controller

gkfla668 committed:
feat: 랜덤 로또숫자 출력

gkfla668 committed:
feat: 지난주 당첨번호 입력 저장

gkfla668 committed:
refactor: getLottoNumber 분리

gch01410 reviewed (changes requested):
MVC 구조 잘 잡고 구현하고 있는데?? ㅋㅋ 코멘트할게 딱히 없네
잘하고 있어. 구현 다하고 리뷰요청 다시 줘!

gkfla668 committed:
feat: 로또 구매 갯수 출력 메세지 메소드 분리

gkfla668 committed:
feat: 보너스 넘버 입력 메소드 분리
fix: 사용자 로또 넘버 구하는 메소드 수정

gkfla668 committed:
feat: 당첨 통계를 위한 매치 함수 기능 추가

gkfla668 committed:
refactor: getUserLottoNumber 메소드 분리
- 구매 갯수 반환
- 로또 번호 출력

gkfla668 committed:
refactor: 지난주 당첨번호 입력 메소드 수정

gkfla668 committed:
test: 사용자가 구매한 랜덤 로또 갯수 테스트

gkfla668 committed:
fix: 구입 금액, 보너스 넘버 입력 제한 메소드 추가

gkfla668 committed:
fix: 지난 주 당첨 번호 갯수 6개 제한 메소드 추가

gkfla668 committed:
test: 당첨번호 쉼표 기준으로 split Test

gkfla668 committed:
refactor: 지난주 당첨번호 입력하는 메소드 분리
feat: createWinningLotto 메소드 생성

gkfla668 committed:
test: InputValidatorTest 추가

gkfla668 committed:
feat: 사용자 로또 중 당첨 로또와 일치하는 총 갯수들을 담은 리스트를 구하는 메소드

gkfla668 committed:
feat: HashMap 사용하여 당첨 통계 메소드

gkfla668 committed:
feat: HashMap 초기화

gkfla668 committed:
refactor: HashMap key type 전체 수정

gkfla668 committed:
feat: 수익률 구하는 메소드 추가

gkfla668 committed:
refactor: 당첨 통계 내는 메소드와 총 당첨 금액 구하는 메소드 분리

gkfla668 committed:
refactor: 총 당첨금액 구하는 메소드 수정

gkfla668 committed:
refactor: 총 수익률 구하는 메소드 수정

gkfla668 committed:
refactor: HashMap 초기화 메소드 for문 수정

gkfla668 committed:
refactor: getCountOfMatch 메소드 수정

gkfla668 committed:
feat: 총 수익률 소숫점 2번째 자리까지 출력 기능 추가

gkfla668 committed:
fix: 규칙 4_한 줄에 점을 하나만 찍는다.

gkfla668 committed:
fix: (이펙티브 자바) 규칙64_객체는 인터페이스를 사용해 참조하라.

gkfla668 committed:
test: 사용자가 구매한 로또 갯수 계산 Test

gkfla668 committed:
test: 총 수익률 구하는 Test

gkfla668 committed:
refactor: controller 수정, model 분리

gkfla668 committed:
test: model test code 추가

gkfla668 committed:
refactor: controller 수정

gkfla668 committed:
refactor: model 분리

gkfla668 committed:
test: StatisticsTest 이동

gkfla668 committed:
refactor: 전체 코드 리팩토링

gkfla668 committed:
refactor: 매치 통계 등록하는 메소드 model로 분리

gkfla668 committed:
refactor: mvc 패턴 구조 수정

gkfla668 committed:
refactor: controller 수정

gkfla668 committed:
refactor: controller 수정

gch01410 commented on src/main/java/model/WinningLotto.java:
```diff
@@ -0,0 +1,72 @@
+        String[] winningNumber = new String[6];
+
+        do {
+            String lottoNumber = InputView.inputWinningNumber(); // 지난주 당첨번호 입력
```
Model영역에서 View영역으로 직접 요청하고 있어
Controller에서 View영역으로 요청하고 받은 결과값을 Controller가 Model영역으로 넘겨주도록 해보자!

gch01410 reviewed (changes requested):
테스트 코드에서 컴파일 에러가 발생하고 있는데 리뷰요청하기전에 테스트 코드가 모두 통과하는지 확인해줘!
그리고 같은 클래스내에서만 호출되는 메서드는 모두 private으로 선언해주고!
IntelliJ에서 노란색으로 표시되는 부분은 왜 노란색으로 뜨는지 확인하고 왠만하면 InteliiJ가 추천하는 방식으로 변경해보자!

gkfla668 committed:
test: UserTest 수정

gkfla668 committed:
refactor: 접근제한자 변경

gkfla668 committed:
refactor: 변수 초기화 값 변경, 제거

gkfla668 committed:
refactor: createUserLotto 메소드 기능 분리, 수정

gkfla668 committed:
refactor: printPurchaseAmountMessage 메소드 기능 분리
- feat: 로또 구매 갯수 구하는 메소드 생성

gkfla668 committed:
refactor: 메소드 기능 분리
- 로또 게임 총 결과 출력
- 당첨 통계 출력
- 총 수익률 출력

gkfla668 committed:
feat: 당첨 금액 구하는 메소드 추가
refactor: 로또 게임 결과 출력하는 메소드 수정

gkfla668 committed:
feat: 사용자 구입 로또 리스트 출력하는 메소드 추가

gkfla668 committed:
refactor: 전체 코드 리팩토링

gkfla668 committed:
test: UserTest 추가

gkfla668 committed:
test: LottoTest 추가

gkfla668 commented on src/main/java/model/User.java:
```diff
@@ -0,0 +1,48 @@
+
+    // 구매 금액 입력
+    public static int inputCost() {
+        int cost = LOTTO_TICKET_PRICE;
```
앗 그러게요! do while이니까 상관없구나 
실수.. ㅎㅎㅎ

gkfla668 committed:
test: 테스트명 수정

gkfla668 commented:
mvc 패턴만 신경쓰다가 접근제어자랑 테스트코드는 까먹고있었어요.. ㅎㅎ
확실히 mvc 패턴 구조 신경써가면서 구현하려니까 코드들이 전체적으로 깔끔해지고 하나의 메서드가 하나의 역할만 수행하도록 짜게 되더라구요! 🧐요번에 많이 신경썼어요.,, 😅
이렇게 하는게 맞는건지 잘하고 있는건지 잘 모르겠지만 피드백 수정하고 이에 맞춰 테스트코드 수정하고, 추가해서 모든 테스트 코드 통과하는거 확인하고 푸쉬했습니다!

그리고 구현하면서 궁금한게 생겼는데, 서치해도 정확하게 안나와서 여쭤봐요! 
다른 클래스에 있는 메서드를 사용할 때 사용하려는 메서드를 import하고 클래스명 없이 바로 메서드를 사용하는 방식?과 import하지 않고 인스턴스를 생성해서 메서드를 사용하는 방식 중 뭐가 더 나은건가요?? 
아무래도 import하는게 코드가 더 깔끔할까요?

gkfla668 committed:
refactor: 노란색 박스 코드 수정
feat: 지난주 당첨번호 String[]을 List<Integer>로 변환해주는 메소드 추가

gkfla668 commented:
오호 노란색 박스기능 처음알았어요
불필요한 코드 제거해주고,, 더 나은 코드로 변경할 수 있게 알려주네요 !! 굿 
노란색 박스기능 수정해서 리뷰요청 합니당!

gkfla668 committed:
refactor: 접근제어자 수정
- changeStringArrayToList: public > private

gkfla668 committed:
refactor: mvc 패턴, test code 리팩토링

gkfla668 committed:
refactor: 접근 제어자, static method 리팩토링

gkfla668 commented:
최대한 설명해주신거에 대한 이해를 바탕으로 수정해보았습니당
또 잘못 이해한건 아닌지 모르겠지만,, 확실히 컨트롤러의 역할을 알 것 같아요!

model 영역에서 직접적으로 컨트롤러를 호출해 view를 호출하는 것이 아닌 
model과 view는 가만히 있고... 컨트롤러 그 자체에서 view를 호출해 입력받게 하고, 
여기서 전달받은 값을 바로 또 view 호출해서 출력해주거나
컨트롤러가 model 영역으로 전달해서 원하는 값으로 변환하고 값을 다시 받아서 
컨트롤러가 view 호출해 이 값을 출력해주고!! 

이해하고 나니까 그 전에 제대로 이해 못하고 엉망ㅇ으로 해놓은게 눈에 보이네요 ... 😅
컨트롤러가 진짜 완전 중재자 역할인 것 같아요
모든 권한이 컨트롤러에게 있어서 컨트롤러가 명령하는대로 프로그램이 굴러가는 느낌...? 

static도 import안하고 인스턴스로 호출해서 명시해놓으니까 
mvc 패턴 적용하는데에 파악이 더 쉽고 코드를 읽는데 가독성이 좀 높아진 것 같아요!

아 그리고 느낀건데 model 영역 클래스를 나누는거에 있어서  약간 헷갈리고 어렵더라구요
스플릿하거나, 배열을 리스트로 변환해주거나 그런 메서드들을 분리해줘야 할 것 같기도 하고.. 
지금 정도면 갠찮은건가요??

제대로 이해 못 했을수도.. 아직 많이 부족하겠지만.. 일단 리뷰 요청합니다!
바쁘실텐데 감사해용! ☺️

gkfla668 committed:
feat: model 영역 User Class 추가
- UserTest 작성

gch01410 commented on src/main/java/controller/InputController.java:
```diff
@@ -0,0 +1,45 @@
+public class InputController {
+    // 구입 금액 반환
+    public static int inputCost() {
+        int cost = 0;
```
IntelliJ에서 회색으로 표시되는 부분도 한번 확인해봐!
회색으로 표시되면 불필요하다는 뜻이야!
int형 변수를 초기화안하면 자동으로 0으로 초기화 되거든!

gch01410 commented on src/main/java/model/InputValidator.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    // 구입 금액_단위_1000원 단위인지
+    public static boolean isValidateCostUnit(int cost) {
+        return ((cost % MIN_LOTTO_COST) == LOTTO_COST_REMAINDER);
+    }
+
+    // 구입 금액_최소금액_1000원인지
+    public static boolean isValidateCostMinCost(int cost) {
```
같은 클래스에서만 호출되는 메서드는 private으로 선언해줘 ㅜㅜ
테스트 코드 작성때문이라면 private메서드를 호출하는 public 메서드의 테스트 코드를 통해 검증해보자!

gch01410 commented on src/main/java/controller/InputController.java:
```diff
@@ -0,0 +1,45 @@
+
+        do {
+            String lottoNumber = InputView.inputWinningNumber(); // view 호출, 지난주 당첨번호 입력
+            winningNumber = Lotto.splitNumbers(lottoNumber); // 당첨번호 쉼표표 기준으 split
```
쉼표로 분리하는 로직을 Lotto가 갖고 있어야만할까?
입력된 값을 가공하는 기준을 로또는 모르고 있어도 될 것 같아.
입력된 값이 어떤형식이든지 로또는 Int형 리스트로 로또번호를 받기만 하면 되니깐!

gkfla668 replied:
네 그래서 스플릿 메서드랑 배열을 리스트로 변환해주는 메서드 위치가 애매했었는데 ㅠㅠ
model로 클래스 새로 생성해서 옮겼습니다!

gch01410 commented on src/test/java/model/InputValidatorTest.java:
```diff
@@ -0,0 +1,111 @@
+        int cost = 500;
+
+        //when
+        boolean result = inputValidator.isValidateCostMinCost(cost);
```
노란색 박스 표시되는데 확인해보고 수정헤줘!

gch01410 commented on src/test/java/model/StatisticsTest.java:
```diff
@@ -0,0 +1,21 @@
+
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+
+public class StatisticsTest {
```
Statistics 클래스에는 당첨통계 구하는 로직, 총 당첨금액 반환하는 로직도 있어!
각 로직의 테스트 코드도 작성해보자!

gch01410 commented on src/main/java/model/WinningLotto.java:
```diff
@@ -0,0 +1,18 @@
+        this.bonusNo = bonusNo;
+    }
+
+    public Rank match(Lotto userLotto) {
```
match 메서드에 대한 테스트 코드도 작성해볼 수 있을 것 같아!

gch01410 commented on src/main/java/model/Lotto.java:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    // 당첨 번호와 사용자 로또 번호 중 일치하는 번호가 몇개인지 반환
+    public int getMatchCount(Lotto winningLotto) {
```
getMatchCount 메서드의 테스트 코드도 작성해볼 수 있을 것 같아!

gch01410 reviewed (changes requested):
MVC 구조 잘 적용했어! 💯 💯 
테스트 코드를 추가할곳이 몇개 보여서 코멘트 남겼어!
조금만 더 힘내보자! 💪

gkfla668 committed:
feat: Convert 클래스 추가
test: ConvertTest, WinningLottoTest 클래스 추가
- Lotto, Statistics, Winning Test code 추가

gkfla668 committed:
test: test code 리팩토링

gch01410 commented on src/main/java/util/RandomLottoNumber.java:
```diff
@@ -0,0 +1,38 @@
+        List<Integer> randomNumbers = new ArrayList<>();
+
+        for (int i = 0; i < LOTTO_TICKET_LENGTH; i++) {
+            int numbers = RandomLottoNumber.getRandomNumber(); // 랜덤 로또 번호 생성
```
같은 클래스내니까 getRandomNumber() 으로만 호출해도 될 것 같아

gch01410 reviewed (approved):
처음으로 MVC 적용 하는것일텐데 잘 이해하고 적용했어!
수고했어!!
